### PR TITLE
[SPARK-19759][ML] not using blas in ALSModel.predict for optimization

### PR DESCRIPTION
**Pull request by mgaido91**
## What changes were proposed in this pull request?

In `ALS.predict` currently we are using `blas.sdot` function to perform a dot product on two `Seq`s. It turns out that this is not the most efficient way.

I used the following code to compare the implementations:

```
def time[R](block: => R): Unit = {
    val t0 = System.nanoTime()
    block
    val t1 = System.nanoTime()
    println("Elapsed time: " + (t1 - t0) + "ns")
}
val r = new scala.util.Random(100)
val input = (1 to 500000).map(_ => (1 to 100).map(_ => r.nextFloat).toSeq)
def f(a:Seq[Float], b:Seq[Float]): Float = {
    var r = 0.0f
    for(i <- 0 until a.length) {
        r+=a(i)*b(i)
    }
    r
}
import com.github.fommil.netlib.BLAS.{getInstance => blas}
val b = (1 to 100).map(_ => r.nextFloat).toSeq
time { input.foreach(a=>blas.sdot(100, a.toArray, 1, b.toArray, 1)) }
// on average it takes 2968718815 ns
time { input.foreach(a=>f(a,b)) }
// on average it takes 515510185 ns
``` 

Thus this PR proposes the old-style for loop implementation for performance reasons.

## How was this patch tested?

existing UTs

### CODE REVIEW

**Review by srowen — CHANGES_REQUESTED** [review]
Yes, seems that people have found often in the past that level 1 BLAS ops were not actually faster. Here using it requires overhead of converting to an array. I can believe this is faster. Usually a while loop is a little faster. 

Interestingly there is another PR suggesting that Level 1 BLAS is faster when native libs are available.

File: mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala [hunk 1]
```diff
@@ -289,9 +289,11 @@ class ALSModel private[ml] (
 
   private val predict = udf { (featuresA: Seq[Float], featuresB: Seq[Float]) =>
     if (featuresA != null && featuresB != null) {
-      // TODO(SPARK-19759): try dot-producting on Seqs or another non-converted type for
-      // potential optimization.
-      blas.sdot(rank, featuresA.toArray, 1, featuresB.toArray, 1)
+      var dotProduct = 0.0f
+      for(i <- 0 until rank) {
```
Comment (WeichenXu123) [inline mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala]: You should `while` instead of `for`

**Comment by WeichenXu123** [comment]
Have you made some test to check the performance difference for this ?

**Comment by mgaido91** [comment]
@srowen I tried enabling native BLAS, but native BLAS implementation is still much slower: average on 10 runs is 2529,922753 ms against 515,510185 ms of the for loop. As a reference, I am using a OSX 2.5 GHz Intel Core i7.
What is worth to notice, though, is that I tried to run the same code but performing the `toArray` before, thus excluding its time from the computation. In this case, native BLAS implementation is much faster: 100,969697 ms. Thus here the "performance killer" is the conversion to array, as you pointed out.

@WeichenXu123 In the description od the PR and here you can see the tests I made. Do you think something else is needed?

**Review by srowen — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master

**Pull request closed without merging**